design_docs/oscode/oscode_sundials.md

```diff
@@ -0,0 +1,347 @@
+# Proposal For Including OSCODE Into Sundials
+
+## Summary
+
+This design document describes the motivation and architecture for adding the OSCODE numerical routine for the efficient solution of second order, linear, homogeneous Ordinary Differential Equations (ODEs) with rapidly oscillating solutions. The proposed solver will will operate with initial value problems (IVPs) of the form
+
+$$ 
+u(t_0) = u_0, \quad u(t_1) = u_1,
+$$
+$$
+\ddot{u}(t) + 2 \gamma(t) \dot{u}(t) + \omega^2(t)u(t) = 0, \quad t \in [t_0, t_1], 
+$$
+
+where $\gamma$ and $\omega$ may or may not be expressed as a closed form function of time. We further require $\gamma$ and $\omega$ to be real-valued and $\omega > 0$.. If $\omega$ is large, the components may become highly oscillatory, and standard (polynomial-based) numerical methods will require $\mathcal{O}(\omega)$ timesteps/discretization points. In regimes where $\omega$ is large and smooth, OSCODE exploits an asympotic approximation to reduce this computational cost to $\mathcal{O}(1)$ ($\omega$-independent). In other regimes of the solution interval, OSCODE behaves as a Runge--Kutta solver, and is thus robust to changes in the behavior in the solution from oscillatory to non-oscillatory.
+
+<!--- We should cite the oscode papers for details, and the JOSS paper! --->
+
+## Motivation
+
+Solutions to oscillatory problems are often bespoke and use analytic approximations such as the Wentzel-Kramers-Brillouin approximation (WKB) method used by OSCODE. Giving researchers access to a general purpose solver for solutions for ODEs of the form written above will allow them to explore more computationally complex models in their research and speed up existing implementations.
+
+The second order ODE can be reposed as a set of two, first order ODEs to fit into the standard schema for sundials solvers,
+
+$$
+\dot{\mathbf{y}} = \begin{pmatrix}
+\mathbf{y}[1]\\
+-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
+\end{pmatrix}
+$$
+
+Among the SUNDIALS modules, ARKODE ---which solves equations with known explicit and implicit, stiff and nonstill timescale components--- is structured most similarly to OSCODE.
+
+$$
+M(t)\dot{\mathbf{y}} = f^E(t, \mathbf{y}) + f^I(t, \mathbf{y}), \quad \mathbf{y}(t_0) = \mathbf{y}_0
+$$
+
+For the new OSCODE module we propose the right-hand-side will be a subset of what ARKODE can handle, specifically,
+
+$$
+M(t)\dot{\mathbf{y}} = f^O(t, \mathbf{y}), \quad \mathbf{y}|(t_0) = \mathbf{y}_0, \\
+$$
+where $f^O(t)$ takes the specific form
+$$
+f^O(t, y) = \begin{pmatrix}
+\mathbf{y}[1]\\
+-\mathbf{y}[0]\omega^2(t)-2\gamma(t)\mathbf{y}[1]
+\end{pmatrix}.
+$$
+
+As the solution $\mathbf{y}(t)$ may vary between oscillatory and smoothly varying, OSCODE will dynamically switch between a Runge-Kutta solver and its asymptotic ARDC-based solver, at each step choosing the method which yields the larger step length (while keeping the local error within user defined tolerance).
```

Comment:
@fruzsinaagocs After this section can you write a blurb describing the ARDC method?